### PR TITLE
PORT-7273 |  DORA Metrics

### PR DESCRIPTION
**Pull request by mk-armah**
# Description
This docs provides a comprehensive guide for creating a GitHub action computes DORA Metrics for a service (repo)

## Inputs
| Name                 | Description                                                                                          | Required | Default            |
|----------------------|------------------------------------------------------------------------------------------------------|----------|--------------------|
| repository              | GitHub repository in owner/repo format eg. port-labs/self-service-actions                               | true     | -                  |
| timeframe              | Time frame in weeks to calculate metrics on                                                              | true    | -               |
| workflow              | The name of the workflows to process. Multiple workflows can be separated by a comma (,) .                                                              | true    | -               |

- Resource Path (`/create-self-service-experiences/setup-backend/github-workflow/examples/dora-metrics`)

### CODE REVIEW

File: docs/create-self-service-experiences/setup-backend/github-workflow/examples/dora-metrics.md [hunk 1]
```diff
@@ -0,0 +1,899 @@
...
+
+  <summary><b>Port Action: Dora Metrics (click to expand)</b></summary>
+   :::tip
+- `<GITHUB-ORG>` - your GitHub organization or user name.
+- `<GITHUB-REPO-NAME>` - your GitHub repository name.
+:::
+
+
+```json showLineNumbers
+[
```
Comment (PeyGis) [inline docs/create-self-service-experiences/setup-backend/github-workflow/examples/dora-metrics.md:130]: Remove the square bracket and the top and end of this JSON

File: docs/create-self-service-experiences/setup-backend/github-workflow/examples/dora-metrics.md [hunk 2]
```diff
@@ -0,0 +1,899 @@
...
+name: Ingest DORA Metrics
+
+on:
+  workflow_dispatch:
+    inputs:
+      repository:
+        description: 'Comma-separated list of repositories to analyze (eg. https://github.com/port-labs/self-service-actions)'
+        required: true
+      timeframe:
+        description: 'Last X weeks'
```
Comment (PeyGis) [inline docs/create-self-service-experiences/setup-backend/github-workflow/examples/dora-metrics.md:204]: Give it a better description

File: docs/create-self-service-experiences/setup-backend/github-workflow/examples/dora-metrics.md [hunk 3]
```diff
@@ -0,0 +1,899 @@
...
+        required: true
+      workflow:
+        description: The name of the workflows to process. Multiple workflows can be separated by a comma (,).
```
Comment (PeyGis) [inline docs/create-self-service-experiences/setup-backend/github-workflow/examples/dora-metrics.md:207]: Why are some `description` in single quote and some don't have quote at all?

File: docs/create-self-service-experiences/setup-backend/github-workflow/examples/dora-metrics.md [hunk 4]
```diff
@@ -0,0 +1,899 @@
...
+          echo "REPOSITORY=$repository_path" >> $GITHUB_ENV
+          echo "TITLE=$title" >> $GITHUB_ENV
+        shell: bash
+
+      - name: Set up Python
+        uses: actions/setup-python@v2
+        with:
+          python-version: '3.x'
+
+      - name: Report Failure In Settting Up Dependencies
```
Comment (PeyGis) [inline docs/create-self-service-experiences/setup-backend/github-workflow/examples/dora-metrics.md:242]: The name of the step can be `Notify intent to install dependencies`

File: docs/create-self-service-experiences/setup-backend/github-workflow/examples/dora-metrics.md [hunk 5]
```diff
@@ -0,0 +1,899 @@
...
+          operation: PATCH_RUN
+          runId: ${{fromJson(github.event.inputs.port_payload).context.runId}}
+          logMessage: "Installing required packages for the action..."
+          
+      - name: Install Python dependencies
+        run: |
+          python -m pip install --upgrade pip
+          pip install -r dora/requirements.txt
+
+      - name: Report Failure In Computing PR Metrics
```
Comment (PeyGis) [inline docs/create-self-service-experiences/setup-backend/github-workflow/examples/dora-metrics.md:258]: The name of the step should be `Report Failure In Settting Up Dependencies`

File: docs/create-self-service-experiences/setup-backend/github-workflow/examples/dora-metrics.md [hunk 6]
```diff
@@ -0,0 +1,899 @@
...
+
+class RepositoryMetrics:
+    def __init__(self, repo_name, time_frame):
+        self.github_client = Github(os.getenv("GITHUB_TOKEN"))
+        self.repo_name = repo_name
+        self.time_frame = int(time_frame)
+        self.start_date = datetime.utcnow().replace(tzinfo=timezone.utc) - timedelta(
+            days=self.time_frame
+        )
+        self.repo = self.github_client.get_repo(f"{self.repo_name}")
```
Comment (PeyGis) [inline docs/create-self-service-experiences/setup-backend/github-workflow/examples/dora-metrics.md:485]: why can't you do this? 
self.github_client.get_repo(self.repo_name)

File: docs/create-self-service-experiences/setup-backend/github-workflow/examples/dora-metrics.md [hunk 7]
```diff
@@ -0,0 +1,899 @@
...
+```
+</details>
+
+<details>
+  <summary><b>Deployment Frequency</b></summary>
+
+```python showLineNumbers title="deploymentfrequency.py"
+from github import Github
+import datetime
+import pytz  # Make sure to install pytz if you haven't: pip install pytz
```
Comment (PeyGis) [inline docs/create-self-service-experiences/setup-backend/github-workflow/examples/dora-metrics.md:637]: If pytz is a required dependency then add it to the requirements.txt

File: docs/create-self-service-experiences/setup-backend/github-workflow/examples/dora-metrics.md [hunk 8]
```diff
@@ -0,0 +1,899 @@
...
+    time_frame = int(os.getenv('TIMEFRAME_IN_DAYS'))
+    number_of_days = 30 if not time_frame else time_frame
+    
+    report = main(owner_repo, workflows, branch, number_of_days)
+    with open(os.getenv('GITHUB_ENV'), 'a') as github_env:
+        github_env.write(f"lead_time_for_changes_report={report}\n")
+```
+</details>
+
+6. Trigger the actions from the [self-service](https://app.getport.io/self-serve) page of your Port application.
```
Comment (PeyGis) [inline docs/create-self-service-experiences/setup-backend/github-workflow/examples/dora-metrics.md:899]: Trigger the action not `actions`

**Review by PeyGis — CHANGES_REQUESTED** [review]

File: docs/create-self-service-experiences/setup-backend/github-workflow/examples/dora-metrics.md (hunk 7 above)
Comment (mk-armah) [reply docs/create-self-service-experiences/setup-backend/github-workflow/examples/dora-metrics.md:635]: Its in the dependencies

File: docs/create-self-service-experiences/setup-backend/github-workflow/examples/dora-metrics.md [hunk 9]
```diff
@@ -0,0 +1,897 @@
+---
+sidebar_position: 10
+---
+
+import PortTooltip from "/src/components/tooltip/tooltip.jsx";
+
+# Dora Metrics
+
+In this guide, we will create a self-service action in Port that executes a GitHub workflow to compute the dora metrics for a service.
```
Comment (MPTG94) [inline docs/create-self-service-experiences/setup-backend/github-workflow/examples/dora-metrics.md:9]:
```suggestion
In this guide, we will create a self-service action in Port that executes a GitHub workflow to compute the DORA metrics for a service.

```

File: docs/create-self-service-experiences/setup-backend/github-workflow/examples/dora-metrics.md [hunk 10]
```diff
@@ -0,0 +1,897 @@
+---
+sidebar_position: 10
+---
+
+import PortTooltip from "/src/components/tooltip/tooltip.jsx";
+
+# Dora Metrics
```
Comment (MPTG94) [inline docs/create-self-service-experiences/setup-backend/github-workflow/examples/dora-metrics.md:7]:
```suggestion
# DORA Metrics

```

File: docs/create-self-service-experiences/setup-backend/github-workflow/examples/dora-metrics.md [hunk 11]
```diff
@@ -0,0 +1,897 @@
...
+</details>
+
+## Create Github workflow
+
+Follow these steps to get started:
+
+1. Create the following GitHub Action secrets:
+    - `PORT_CLIENT_ID` - Port Client ID [learn more](/build-your-software-catalog/custom-integration/api/#get-api-token)
+    - `PORT_CLIENT_SECRET` - Port Client Secret [learn more](/build-your-software-catalog/custom-integration/api/#get-api-token)
+    - `PATTOKEN` - GitHub PAT token. Ensure that Read access to actions and metadata permission is set. [learn more](https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/managing-your-personal-access-tokens)
```
Comment (MPTG94) [inline docs/create-self-service-experiences/setup-backend/github-workflow/examples/dora-metrics.md:117]:
```suggestion
    - `PATTOKEN` - GitHub PAT fine-grained token. Ensure that read-only access to actions and metadata permission is set. [learn more](https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/managing-your-personal-access-tokens#creating-a-fine-grained-personal-access-token)

```

Comment (MPTG94) [reply docs/create-self-service-experiences/setup-backend/github-workflow/examples/dora-metrics.md:117]: I understand from the mentioned permissions that the token needs to be a fine-grained one. If so, the user needs more information such as - to which repositories should he give the token access

File: docs/create-self-service-experiences/setup-backend/github-workflow/examples/dora-metrics.md [hunk 12]
```diff
@@ -0,0 +1,897 @@
...
+    "type": "GITHUB",
+    "org": "<GITHUB-ORG>",
+    "repo": "<GITHUB-REPO-NAME>",
+    "workflow": "dora-metrics.yaml",
+    "omitUserInputs": false,
+    "omitPayload": false,
+    "reportWorkflowStatus": true
+  },
+  "trigger": "DAY-2",
+  "description": "Estimate dora metrics for a service",
```
Comment (MPTG94) [inline docs/create-self-service-experiences/setup-backend/github-workflow/examples/dora-metrics.md:179]:
```suggestion
  "description": "Calculate DORA metrics for a service",

```

File: docs/create-self-service-experiences/setup-backend/github-workflow/examples/dora-metrics.md [hunk 13]
```diff
@@ -0,0 +1,897 @@
...
+   :::tip
+- `<GITHUB-ORG>` - your GitHub organization or user name.
+- `<GITHUB-REPO-NAME>` - your GitHub repository name.
+:::
+
+
+```json showLineNumbers
+{
+  "identifier": "dora_metrics",
+  "title": "Dora Metrics",
```
Comment (MPTG94) [inline docs/create-self-service-experiences/setup-backend/github-workflow/examples/dora-metrics.md:132]:
```suggestion
  "title": "DORA Metrics",

```

File: docs/create-self-service-experiences/setup-backend/github-workflow/examples/dora-metrics.md [hunk 14]
```diff
@@ -0,0 +1,897 @@
...
+          general context (blueprint, run id, etc...)'
+        type: string
+      
+jobs:
+  compute-dora-metrics:
+    runs-on: ubuntu-latest
+    steps:
+    
+      - name: Checkout code
+        uses: actions/checkout@v2
```
Comment (MPTG94) [inline docs/create-self-service-experiences/setup-backend/github-workflow/examples/dora-metrics.md:219]:
```suggestion
        uses: actions/checkout@v4

```

File: docs/create-self-service-experiences/setup-backend/github-workflow/examples/dora-metrics.md [hunk 15]
```diff
@@ -0,0 +1,897 @@
...
+          title=$(echo "$cleaned_name" | awk '{for(i=1;i<=NF;i++) $i=toupper(substr($i,1,1)) tolower(substr($i,2));}1')
+
+          # export to github enviroment vars
+          echo "TIMEFRAME_IN_DAYS=$days" >> $GITHUB_ENV
+          echo "REPOSITORY=$repository_path" >> $GITHUB_ENV
+          echo "TITLE=$title" >> $GITHUB_ENV
+        shell: bash
+
+      - name: Set up Python
+        uses: actions/setup-python@v2
```
Comment (MPTG94) [inline docs/create-self-service-experiences/setup-backend/github-workflow/examples/dora-metrics.md:236]:
```suggestion
        uses: actions/setup-python@v5

```

File: docs/create-self-service-experiences/setup-backend/github-workflow/examples/dora-metrics.md [hunk 16]
```diff
@@ -0,0 +1,897 @@
...
+    
+    with open(os.getenv('GITHUB_ENV'), 'a') as github_env:
+        github_env.write(f"deployment_frequency_report={report}\n")
+```
+</details>
+
+<details>
+  <summary><b>Lead Time For Changes</b></summary>
+
+```python showLineNumbers title="leadtimeforchanges.py"
```
Comment (MPTG94) [inline docs/create-self-service-experiences/setup-backend/github-workflow/examples/dora-metrics.md:729]:
```suggestion
```python showLineNumbers title="lead_time_for_changes.py"

```

File: docs/create-self-service-experiences/setup-backend/github-workflow/examples/dora-metrics.md [hunk 17]
```diff
@@ -0,0 +1,897 @@
...
+
+```text showLineNumbers title="requirements.txt"
+PyGithub==2.3.0
+requests==2.31.0
+pytz==2024.1
+```
+
+</details>
+
+5. Create the following python scripts (`calculate_pr_metrics.py`, `deploymentfrequency.py` and `leadtimeforchanges.py` ) in a folder named `dora` at the root of your GitHub repository:
```
Comment (MPTG94) [inline docs/create-self-service-experiences/setup-backend/github-workflow/examples/dora-metrics.md:462]:
```suggestion
5. Create the following python scripts (`calculate_pr_metrics.py`, `deployment_frequency.py` and `lead_time_for_changes.py` ) in a folder named `dora` at the root of your GitHub repository:

```

File: docs/create-self-service-experiences/setup-backend/github-workflow/examples/dora-metrics.md [hunk 18]
```diff
@@ -0,0 +1,897 @@
...
+          runId: ${{fromJson(github.event.inputs.port_payload).context.runId}}
+          logMessage: "Computing lead time for changes metric ..."
+          
+      - name: Lead Time For Changes
+        id: lead_time_for_changes
+        env:
+          WORKFLOWS: ${{ inputs.workflow }}
+          GITHUB_TOKEN: ${{ secrets.PATTOKEN }} 
+          REPOSITORY: ${{ env.REPOSITORY }}
+        run: python dora/leadtimeforchanges.py
```
Comment (MPTG94) [inline docs/create-self-service-experiences/setup-backend/github-workflow/examples/dora-metrics.md:359]:
```suggestion
        run: python dora/lead_time_for_changes.py

```

File: docs/create-self-service-experiences/setup-backend/github-workflow/examples/dora-metrics.md [hunk 19]
```diff
@@ -0,0 +1,897 @@
...
+          runId: ${{fromJson(github.event.inputs.port_payload).context.runId}}
+          logMessage: "Computing deployment frequency metric ..."
+          
+      - name: Deployment Frequency
+        id: deployment_fequency
+        env:
+          WORKFLOWS: ${{ inputs.workflow }}
+          GITHUB_TOKEN: ${{ secrets.PATTOKEN }} 
+          REPOSITORY: ${{ env.REPOSITORY }}
+        run: python dora/deploymentfrequency.py
```
Comment (MPTG94) [inline docs/create-self-service-experiences/setup-backend/github-workflow/examples/dora-metrics.md:320]:
```suggestion
        run: python dora/deployment_frequency.py

```

File: docs/create-self-service-experiences/setup-backend/github-workflow/examples/dora-metrics.md [hunk 20]
```diff
@@ -0,0 +1,897 @@
...
+
+if __name__ == "__main__":
+    main()
+```
+</details>
+
+<details>
+  <summary><b>Deployment Frequency</b></summary>
+
+```python showLineNumbers title="deploymentfrequency.py"
```
Comment (MPTG94) [inline docs/create-self-service-experiences/setup-backend/github-workflow/examples/dora-metrics.md:632]:
```suggestion
```python showLineNumbers title="deployment_frequency.py"

```

File: docs/create-self-service-experiences/setup-backend/github-workflow/examples/dora-metrics.md [hunk 21]
```diff
@@ -0,0 +1,897 @@
...
+    time_frame = int(os.getenv('TIMEFRAME_IN_DAYS'))
+    number_of_days = 30 if not time_frame else time_frame
+    
+    report = main(owner_repo, workflows, branch, number_of_days)
+    with open(os.getenv('GITHUB_ENV'), 'a') as github_env:
+        github_env.write(f"lead_time_for_changes_report={report}\n")
+```
+</details>
+
+6. Trigger the action from the [self-service](https://app.getport.io/self-serve) page of your Port application.
```
Comment (MPTG94) [inline docs/create-self-service-experiences/setup-backend/github-workflow/examples/dora-metrics.md:897]: Please add in the doc an example of inputs to trigger the action with, as I believe users will have a hard time understanding what are the correct inputs for the `workflow` and `repository` inputs in the action form

**Review by MPTG94 — CHANGES_REQUESTED** [review]
Please make changes according to my review

**Pull request closed without merging**